crates/l2/sequencer/prover_server.rs

```diff
@@ -174,72 +165,72 @@ impl ProverServer {
+            if let Err(e) = shutdown_tx.send(()).await {
+                error!("Error sending shutdown message through the mpsc::channel {e}");
+            };
+        });
 
-    pub async fn start(&mut self, rx: Receiver<()>) -> Result<(), ProverServerError> {
         let listener = TcpListener::bind(format!("{}:{}", self.ip, self.port)).await?;
 
-        info!("Starting TCP server at {}:{}", self.ip, self.port);
+        let concurent_clients = 3;
+        let sem = tokio::sync::Semaphore::new(concurent_clients);
+        info!(
+            "Starting TCP server at {}:{} accepting {concurent_clients} concurrent clients.",
+            self.ip, self.port
+        );
 
         loop {
-            match listener.accept().await {
-                Ok((stream, _)) => {
-                    debug!("Connection established!");
-
-                    if let Ok(()) = rx.try_recv() {
-                        info!("Shutting down Prover Server");
-                        break;
-                    }
-
-                    if let Err(e) = self.handle_connection(stream).await {
-                        error!("Error handling connection: {}", e);
-                    }
+            tokio::select! {
+                _ = shutdown_rx.recv() => {
+                    debug!("Shutting down...");
+                    // It will return from the main_logic() with an `Ok(())`
+                    // And inside the run() function the loop will break
+                    // and the prover_server task will finish gracefully.
+                    break;
                 }
-                Err(e) => {
-                    error!("Failed to accept connection: {}", e);
+                res = listener.accept() => {
+                    match res {
+                        Ok((stream, addr)) => {
+                            // Try to acquire a semaphore slot
+                            if let Ok(_guard) = sem.try_acquire() {
+                                debug!("Connection established!");
+                                // Cloning the ProverServer structure to use the handle_connection() fn
+                                // in every spawned task.
+                                // The important fields are `Store` and `EthClient`
+                                // Both fields are wrapped with an Arc, making it possible to clone
+                                // the entire structure.
+                                let mut self_clone = self.clone();
+                                tokio::task::spawn(async move {
+                                    if let Err(e) = self_clone.handle_connection(stream).await {
+                                        error!("Error handling connection from {addr}: {e}");
+                                    } else {
+                                        debug!("Connection from {addr} handled successfully");
+                                    }
+                                });
+
+                            } else {
```

Comment:
We may want to hold a handle here to wait for tasks. Also, we need to move the semaphore guard inside the the task for the semaphore to actually limit how many we fire up:
```suggestion
                                tokio::task::spawn(async move {
                                    let _guard = _guard;
                                    if let Err(e) = self_clone.handle_connection(stream).await {
                                        error!("Error handling connection from {addr}: {e}");
                                    } else {
                                        debug!("Connection from {addr} handled successfully");
                                    }
                                });

                            } else {
```
Otherwise, we drop it at the end of the `if` rather than the end of the task.